sandbox/apparmor/notify/listener/listener.go

```diff
@@ -0,0 +1,251 @@
+//
+// If the kernel does not support the notification mechanism the error is ErrNotSupported.
+func Register() (*Listener, error) {
+	path := notify.SysPath
+	if override := os.Getenv("PROMPT_NOTIFY_PATH"); override != "" {
+		path = override
+	}
+
+	notifyFile, err := osOpen(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, ErrListenerNotSupported
+		}
+		return nil, fmt.Errorf("cannot open %q: %v", path, err)
+	}
+
+	msg := notify.MsgNotificationFilter{ModeSet: notify.APPARMOR_MODESET_USER}
+	data, err := msg.MarshalBinary()
+	if err != nil {
+		notifyFile.Close()
+		return nil, err
+	}
+	ioctlBuf := notify.BytesToIoctlRequestBuffer(data)
+	_, err = notifyIoctl(notifyFile.Fd(), notify.APPARMOR_NOTIF_SET_FILTER, ioctlBuf)
+	if err != nil {
+		notifyFile.Close()
+		return nil, fmt.Errorf("cannot notify ioctl to modeset user on %q: %v", path, err)
+	}
+
+	poll, err := epoll.Open()
+	if err != nil {
+		notifyFile.Close()
+		return nil, fmt.Errorf("cannot open epoll file descriptor: %v", err)
+	}
+	// XXX: Do we need a notification for Writable, to send responses back?
+	if err := poll.Register(int(notifyFile.Fd()), epoll.Readable); err != nil {
+		notifyFile.Close()
+		poll.Close()
+		return nil, fmt.Errorf("cannot register epoll on %q: %v", path, err)
+	}
+
+	listener := &Listener{
+		R: make(chan *Request),
+		E: make(chan error),
+
+		notifyFile: notifyFile,
+		poll:       poll,
+	}
+	return listener, nil
+}
+
+func (l *Listener) decodeAndDispatchRequest(buf []byte, tomb *tomb.Tomb) error {
+	var nmsg notify.MsgNotification
+	if err := nmsg.UnmarshalBinary(buf); err != nil {
+		return err
+	}
+	// What kind of notification message did we get?
+	switch nmsg.NotificationType {
+	case notify.APPARMOR_NOTIF_OP:
+		var omsg notify.MsgNotificationOp
+		if err := omsg.UnmarshalBinary(buf); err != nil {
+			return err
+		}
+		// What kind of operation notification did we get?
+		switch omsg.Class {
+		case notify.AA_CLASS_FILE:
+			var fmsg notify.MsgNotificationFile
+			if err := fmsg.UnmarshalBinary(buf); err != nil {
+				return err
+			}
+			// log.Printf("notification request: %#v\n", fmsg)
+			req := newRequest(l, &fmsg)
+			l.R <- req
+			tomb.Go(func() error {
+				l.waitAndRespond(req, &fmsg)
+				return nil
+			})
+		default:
+			return fmt.Errorf("unsupported mediation class: %v", omsg.Class)
+		}
+	default:
+		return fmt.Errorf("unsupported notification type: %v", nmsg.NotificationType)
+	}
+	return nil
+}
+
+func (l *Listener) waitAndRespond(req *Request, msg *notify.MsgNotificationFile) {
+	resp := notify.ResponseForRequest(&msg.MsgNotification)
+	resp.MsgNotification.Error = 0 // ignored in responses
+	resp.MsgNotification.NoCache = 1
+	if allow := <-req.YesNo; allow {
+		resp.Allow = msg.Allow | msg.Deny
+		resp.Deny = 0
+		resp.Error = 0
+	} else {
+		resp.Allow = 0
+		resp.Deny = msg.Deny
+		resp.Error = msg.Error
+		// msg.Error is field from MsgNotificationResponse, and is unused.
```

Comment:
I added the comment because I was very confused by the several layers of embedded structs having `Error` fields and not knowing how to differentiate them. I think it's possible the error fields could become useful in future versions of the messaging protocol, so I was envisioning this comment (and that on the line directly below) as a cairn of sort to future me or others who might have the misfortune of adjusting the snapd side of the protocol.